workers/unity/Assets/Playground/Scripts/EntityInitialization/EntityGameObjectCreator.cs

```diff
@@ -1,46 +1,69 @@
 
-        public EntityGameObjectCreator(World world, Dictionary<string, GameObject> initialCachedPrefabs = null)
+        public GameObject CreateGameObjectForEntity(SpatialOSEntity entity, Worker worker)
         {
-            this.world = world;
-            cachedPrefabs = initialCachedPrefabs ?? new Dictionary<string, GameObject>();
-        }
+            if (!cachedPrefabsPerWorkerType.TryGetValue(worker.WorkerType, out var cachedPrefabs))
+            {
+                cachedPrefabsPerWorkerType[worker.WorkerType] = cachedPrefabs = new Dictionary<string, GameObject>();
+            }
 
-        public GameObject CreateEntityGameObject(Entity entity, string prefabPath, Vector3 position,
-            Quaternion rotation, EntityId spatialEntityId)
-        {
-            if (!cachedPrefabs.TryGetValue(prefabPath, out var prefab))
+            if (!entity.HasComponent<SpatialOSMetadata>())
+            {
+                return null;
+            }
+
+            var prefabName = entity.GetComponent<SpatialOSMetadata>().EntityType;
+            if (!entity.HasComponent<SpatialOSTransform>())
+            {
+                return cachedPrefabs[prefabName] = null;
+            }
+
+            var transform = entity.GetComponent<SpatialOSTransform>();
+            var position = new Vector3(transform.Location.X, transform.Location.Y, transform.Location.Z) +
+                worker.Origin;
+            var rotation = new UnityEngine.Quaternion(transform.Rotation.X, transform.Rotation.Y,
+                transform.Rotation.Z, transform.Rotation.W);
+            var workerSpecificPath = Path.Combine("Prefabs", worker.WorkerType, prefabName);
+            var commonPath = Path.Combine("Prefabs", "Common", prefabName);
+
+            if (!cachedPrefabs.TryGetValue(workerSpecificPath, out var prefab)
+                && !cachedPrefabs.TryGetValue(commonPath, out prefab))
             {
-                prefab = Resources.Load<GameObject>(prefabPath);
+                prefab = Resources.Load<GameObject>(workerSpecificPath);
                 if (prefab == null)
                 {
-                    throw new PrefabNotFoundException($"Prefab for prefabPath {prefabPath} not found.");
+                    prefab = Resources.Load<GameObject>(commonPath);
                 }
 
-                cachedPrefabs[prefabPath] = prefab;
+                // Cache even if null
+                cachedPrefabs[prefabName] = prefab;
             }
 
-            var gameObject = GameObject.Instantiate(prefab, position, rotation);
-            gameObject.name = $"{prefab.name}(SpatialOS: {spatialEntityId}, Unity: {entity.Index}/{world.Name})";
-            return gameObject;
-        }
-    }
-
-    public class PrefabNotFoundException : Exception
-    {
-        public PrefabNotFoundException(string message) : base(message)
-        {
+            if (prefab == null)
+            {
+                return null;
+            }
+            else
```

Comment:
no need for the else. You will never get to this part anyway, if the prefab is null.